src/heap/sharedMem.ml

```diff
@@ -1071,6 +1111,298 @@ module NewAPI = struct
+    let l = ref 0 in
+    while !r land 3 == 3 do
+      (* two random bits will be set with probability 1/4 *)
+      incr l;
+      r := !r lsr 2
+    done;
+    !l
+
+  let sklist_size = int_size + (sklist_num_levels * addr_size)
+
+  (* The skip list stores a max search level, which is the highest level of any
+   * node in the set. We start searches at this level instead of the highest
+   * possible level as an optimization. *)
+  let write_sklist chunk =
+    let addr = write_header chunk Sklist_tag sklist_size in
+    unsafe_write_tagged_int chunk 0;
+    for i = 1 to sklist_num_levels do
+      unsafe_write_addr chunk null_addr
+    done;
+    addr
+
+  let prepare_write_sknode data =
+    let level = sklist_random_level () in
+    let size = (2 + level) * addr_size in
+    let write chunk =
+      let addr = write_header chunk Sklist_tag size in
+      unsafe_write_addr chunk data;
+      for i = 0 to level do
+        unsafe_write_addr chunk null_addr
+      done;
+      addr
+    in
+    (size, write)
+
+  let sklist_search_level_addr sklist = addr_offset sklist 1
+
+  let sklist_succ_addr level sklist = addr_offset sklist (2 + level)
+
+  let sklist_get_search_level = get_generic_int sklist_search_level_addr
+
+  let sknode_data_addr sknode = addr_offset sknode 1
+
+  let sknode_succ_addr = sklist_succ_addr
+
+  let sknode_get_data = get_generic sknode_data_addr
+
+  let sknode_max_level sknode =
+    let hd = read_header (get_heap ()) sknode in
+    obj_size hd - 2
+
+  let sklist_is_marked x = Int64.(equal 1L (logand x 1L))
+
+  let sklist_marked x = Int64.logor x 1L
+
+  let sklist_unmark x = Int64.(to_int (logand x (lognot 1L)))
+
+  (* Search the skip list, returning true if the a node with `data` is in the
+   * set.
+   *
+   * This function will physically delete any marked nodes by updating the
+   * node's predecessor's successor pointer to the node's successor.
+   *
+   * The caller-provided `preds` and `succs` arrays are populated with the
+   * target node's predecessors and successors at each level. These arrays are
+   * used to implement `sklist_add` and `sklist_remove`. *)
+  let sklist_find cmp sklist data preds succs =
+    let rec start () =
+      let top_level = sklist_num_levels - 1 in
+      let pred = sklist_succ_addr top_level sklist in
+      let curr = sklist_unmark (load_acquire pred) in
+      search top_level pred curr
+    and search level pred curr =
+      if curr == null_addr then
+        (* We are at the end of the list at this level. *)
+        down level pred curr
+      else
+        let succ_addr = sknode_succ_addr level curr in
+        let succ64 = load_acquire succ_addr in
+        let succ = sklist_unmark succ64 in
+        if sklist_is_marked succ64 then
+          (* `curr` is logically deleted. Try to physically remove `curr` from
+           * this level by updating `pred` to point directly to `succ`. *)
+          if compare_exchange_strong pred (i64 curr) (i64 succ) then
```

Comment:
it seems to depend on if compare_exchange is inlined or not

Reply:
I can run the tests under gdb, set a breakpoint on `hh_compare_exchange`, and observe that callers do not allocate Int64.t boxes here. I'm looking at x86, and the int values are just shifted right to decode from the tagged int representation and stored into the rsi/rdx registers before the call into C.

Once I publish my work that uses this code, we can run it through profilers to see what jumps out. Maybe there are surprise allocations, but it will be much easier to see later, and we can fix it then. Sound good?